File: precision.js
// URL input binding
// This input binding is very similar to textInputBinding from
// shiny.js.
var precisionBinding = new Shiny.InputBinding();

// An input binding must implement these methods
$.extend(precisionBinding, {
	// This returns a jQuery object with the DOM element
  find: function(scope) {
    return $(scope).find('.precision-input input');
  },
  
  // return the ID of the DOM element
  getId: function(el) {
    return el.id;
  },
  
  // Given the DOM element for the input, return the value
  getValue: function(el) {
    return  parseFloat($(el).val().replace(',', '.'));
  },
  
  // Given the DOM element for the input, set the value
  setValue: function(el, value) {
    $(el).val(String(value).replace('.', ','));
  },
  
  //implanta metodo para realizar calculo
  changeVal: function(el, opr){
  	let input = $(el)
  	let num = input.attr('data-num').split(',').map(parseFloat)
	let val = new Decimal(parseFloat(input.val().replace(',', '.')))
	let pos = val.times(Decimal.pow(10, val.dp())).toNumber()
	let newVal
	let i = num.indexOf(pos)
	if(i>=0){
		let l = num.length
		i+=opr
		newVal = val.mul(Decimal.div(num[(i+l)%l],num[i-opr]))
		if(i == -1 || i == l)
			newVal = newVal.mul(Decimal.pow(10, Math.sign(i)))
	} else{
		newVal = val.mul(Decimal.pow(10, opr))
	}
	input.val(newVal.toPrecision().replace('.', ','))
  },
  // Set up the event listeners so that interactions with the
  // input will result in data being sent to server.
  // callback is a function that queues data to be sent to
  // the server.
  subscribe: function(el, callback) {
  	var mtply = x => this.changeVal(el, x)
	
	$(el).on('keydown.precisionBinding', function(e) {
		if (e.keyCode == 38 | e.keyCode == 40) {
		  	mtply(e.keyCode == 38 ? 1 : -1)
		  	callback();
		}
	});
	
	$(el).on('wheel.precisionBinding', function(e) {
		e.preventDefault();
	  	mtply(e.originalEvent.deltaY > 1 ? -1 : 1)
	  	callback();
	});

	$(el).parent().find('span').on('click.precisionBinding', function(e) {
	  mtply(parseFloat($(this).attr('data-opr')));
	  callback();
	});
	
	$(el).on('input.precisionBinding keyup.precisionBinding', function(e){
		$(this).val(this.value.replace(/[^0-9.,]+/,'').replace('.', ','))
		callback(true);
	});
  },
  
  // Remove the event listeners
  unsubscribe: function(el) {
    $(el).off('.precisionBinding');
  },
  
  // Receive messages from the server.
  // Messages sent by updatePrecision() are received by this function.
  receiveMessage: function(el, data) {
    if (data.hasOwnProperty('value'))
      this.setValue(el, data.value);
      
    if (data.hasOwnProperty('label'))
      $(el).parents().find('label[for="' + $(el).attr('id') + '"]').text(data.label);
      
    if(data.hasOwnProperty('nums'))
    	$(el).attr('data-num', data.nums.join(','));

    $(el).trigger('input');
  },
  
  // This returns a full description of the input's state.
  // Note that some inputs may be too complex for a full description of the
  // state to be feasible.
  getState: function(el) {
    return {
      label: $(el).parent().find('label[for="' + $escape(el.id) + '"]').text(),
      value: $(el).val(),
      nums:  $(el).attr('data-nums')
    };
  },
  
  // The input rate limiting policy
  getRatePolicy: function() {
    return {
      // Can be 'debounce' or 'throttle'
      policy: 'debounce',
      delay: 250
    };
  }
});

Shiny.inputBindings.register(precisionBinding, "cesarep.precisionInput");

File: precision.r
precisionInput <- function(inputId, label, value = 0.05, nums=c(1,2,5)){
	tagList(
		shiny::singleton(shiny::tags$head(tags$script(HTML(precisionJS)),
													 shiny::tags$script(src = "https://cdnjs.cloudflare.com/ajax/libs/decimal.js/10.2.0/decimal.min.js")
		)
		),
		shiny::tags$div(class='shiny-input-container form-group precision-input',
							 shiny::tags$label('for' = inputId, label),
							 shiny::tags$div(class='input-group',
							 					 shiny::tags$span(class='input-group-addon btn btn-default', 'data-opr'=1,'for'=inputId, shiny::HTML("&#9650;")),
							 					 shiny::tags$input(class="form-control", id=inputId, type='text', value=gsub("\\.", ",", c(as.character(value))), 'data-num'=paste(nums, collapse = ',') ),
							 					 shiny::tags$span(class='input-group-addon btn btn-default', 'data-opr'=-1, 'for'=inputId, shiny::HTML("&#9660;"))
					)
		)
	)
}

updatePrecision <- function(session, inputId, label = NULL, value = NULL, nums=NULL) {
	message <- list(label = label, value = value, nums = nums)
	print('muda input')
	print(message)
	session$sendInputMessage(inputId, message[!vapply(message, is.null, logical(1))])
}


library(readr)

precisionJS <- read_file('precision.js')

"var precisionBinding = new Shiny.InputBinding();
$.extend(precisionBinding, {
  find: function(scope) {
    return $(scope).find('.precision-input input');
  },
  getValue: function(el) {
    return  parseFloat($(el).val().replace(',', '.'));
  },
  subscribe: function(el, callback) {
  	 $(el).on('change.precisionBinding', function(e) {
      callback();
    });
  },
  unsubscribe: function(el) {
    $(el).off('.precisionBinding');
  }
});

Shiny.inputBindings.register(precisionBinding);

mtply = function(input, opr){
	let val = new Decimal(parseFloat(input.val().replace(',', '.')))
	let newVal = val.mul(Decimal.pow(10, opr)).toPrecision()
	input.val(newVal.replace('.', ','))
}

$(document).on('click', '.precision-input span', function(evt) {
  let input = $('#' + $(evt.target).attr('for'));
  let opr = $(evt.target).attr('data-opr');
  mtply(input, opr)
  input.trigger('change');
});

$(document).on('keydown wheel', '.precision-input input', function(evt) {
   let input = $(evt.target);
  	let delta = evt.originalEvent.deltaY
	if (evt.keyCode == 38 | evt.keyCode == 40) {
	  	mtply(input, (evt.keyCode == 38 ? 1 : -1))
	  	input.trigger('change');
	} else if (isFinite(delta) && delta !== 0) {
		evt.preventDefault();
	  	mtply(input, delta > 1 ? -1 : 1)
	  	input.trigger('change');
	}
});

$(document).on('input', '.precision-input input', function(evt){
	let input = $(evt.target)
	input.val(input.val().replace(/[^0-9.,]+/,''))
})
"